### ISSUE TITLE
Support for inception-type architectures (v3, v4) / pre-trained models

### ISSUE THREAD

**Issue by revilokeb** [beniz/deepdetect#89]
**Inception v3** (http://arxiv.org/abs/1512.00567) has been published at the same time as residual nets and show a comparable classification accuracy.
**Inception v4** and InceptionResNet (http://arxiv.org/abs/1602.07261) show very attractive classification accuracy at moderate computational cost, in particular the introduction of residual connections seems to accelerate training significantly.

Given the above properties it is desirable to make those architectures together with ImageNet pre-trained models available in deepdetect.

As of today I am not aware of any publicly available implementation of the **inception v4** or inception_resnet architecture and / or ImageNet pre-trained model.

**Inception v3** however is currently available from the following sources:
- Tensorflow: both model weights and architecture (https://github.com/tensorflow/models/tree/master/inception, https://github.com/tensorflow/tensorflow/blob/master/tensorflow/g3doc/how_tos/image_retraining/index.md)
- dmlc/mxnet: model architecture and weights converted from tensorflow (https://github.com/dmlc/mxnet-model-gallery/blob/master/imagenet-1k-inception-v3.md)
- Torch: model architecture and weights converted from tensorflow (https://github.com/Moodstocks/inception-v3.torch)
- Caffe: model architecture derived from mxnet, ongoing effort to learn the weights (https://github.com/smichalowski/google_inception_v3_for_caffe)

The most reasonable approach to make **inception v3** available in deepdetect thus seems to be the following:
- evaluate caffe model architecture provided by https://github.com/smichalowski/google_inception_v3_for_caffe
- convert weights from tensorflow
- compare predictions to original tensorflow model

**Inception v4 / InceptionResNet** seems to require building / learning from scratch.

I would then start looking into the inception v3 part.

EDIT by @beniz (11/04/2016):

All inception models now supported, partly via Tensorflow backend support, see #209 

|                          | Caffe | Tensorflow | Source        | Top-1 Accuracy (ImageNet) |                                                                                                                                                                                   
|--------------------------|-------|------------|---------------|---------------------------|                                                                                                                                                                                   
| AlexNet                  | Y     | N          | BVLC          |          57.1%                 |                                                                                                                                                                              
| SqueezeNet               | Y     | N          | DeepScale              |       59.5%                    |                                                                                                                                                                     
| Inception v1 / GoogleNet | [Y](https://deepdetect.com/models/ggnet/bvlc_googlenet.caffemodel)     | [Y](https://deepdetect.com/models/tf/inception_v1.pb)          | BVLC / Google |             67.9%              |                                                         
| Inception v2             | N     | [Y](https://deepdetect.com/models/tf/inception_v2.pb)          | Google        |     72.2%                      |                                                                                                                          
| Inception v3             | N     | [Y](https://deepdetect.com/models/tf/inception_v3.pb)          | Google        |         76.9%                  |                                                                                                                          
| ResNet 50                | [Y](https://deepdetect.com/models/resnet/ResNet-50-model.caffemodel)     | [Y](https://deepdetect.com/models/tf/resnet_v1_50/resnet_v1_50.pb)          | MSR           |      75.3%                     |                                          
| ResNet 101               | [Y](https://deepdetect.com/models/resnet/ResNet-101-model.caffemodel)     | [Y](https://deepdetect.com/models/tf/resnet_v1_101/resnet_v1_101.pb)          | MSR           |        76.4%                   |                                       
| ResNet 152               | [Y](https://deepdetect.com/models/resnet/ResNet-152-model.caffemodel)     | [Y](https://deepdetect.com/models/tf/resnet_v1_50/resnet_v1_152.pb)         | MSR           |               77%            |                                           
| Inception-ResNet-v2      | N     | [Y](https://deepdetect.com/models/tf/inception_resnet_v2.pb)          | Google        |       79.79%                    |                                                                                                                  
| VGG-16                   | [Y](https://deepdetect.com/models/vgg_16/VGG_ILSVRC_16_layers.caffemodel)     | [Y](https://deepdetect.com/models/tf/vgg_16/vgg_16.pb)          | Oxford        |               70.5%            |                                                 
| VGG-19                   | [Y](https://deepdetect.com/models/vgg_16/VGG_ILSVRC_16_layers.caffemodel)     | [Y](https://deepdetect.com/models/tf/vgg_19/vgg_19.pb)          | Oxford        |               71.3%            |

**Comment by beniz** [comment]
Agreed that v3 can be the first step, and would be useful to may. v4 would be great to have, learning from scratch would be time-consuming but `inception-resnet-v1` and `inception-resnet-v2` do appear to converge faster (half from Figure 21 ?) than the non-resnet counterparts, so it should be manageable. One good way to start with could be to build a network for 128x128 or even lower image size and get a feel of the convergence rate.

Also note there's a current effort to bring Tensorflow into dd, starting with inception v3, see https://github.com/beniz/deepdetect/issues/30#issuecomment-198547372

My hunch at this stage is that it'd still be worth trying to import the TF weights into a Caffe model as you are suggesting. It'll be some time until a TF integration makes it into stable dd.

Let me know how we can help. At minima, we can report on testing new architectures and weights. For testing purposes, note the new predict call with measures https://github.com/beniz/deepdetect/pull/86 that now allows to test over a repository of images outside of the training call. So comparison should be pretty straightforward to setup.

**Comment by revilokeb** [comment]
Current (not so succesfull) state of that issue based on my experiments so far can be found here: https://github.com/soeaver/caffe-model/issues/1

**Comment by beniz** [comment]
Can you share your setup, along with prototxt files, and possibly the lmdb as well ? Though we could work the latter out. We have additional computational capabilities and we may be able to schedule and resume partial runs in between other activities.

**Comment by revilokeb** [comment]
@beniz sure, I can / will share solver / train_val / caffemodel etc over the next days. LMDB is a bit more problematic due to sheer size and copyright, see ImageNet (I have currently access through affiliation with a University; if you already have ILSVRC12 data I can of course make my LMDB available to you on some shared space, let me know). 

Anyway I seem to have found at least one factor in my solver that has slowed down / hindered learning, see again https://github.com/soeaver/caffe-model/issues/1, so hopefully second try will be more successfull :-)

**Comment by beniz** [comment]
Yes, I have access to to ILSVRC12 data, no worries. I've trained the full GoogleNet before (v1). I'd recommend starting on CIFAR100 first, or ILSVRC12 with smaller images. Also, SGD with fixed lr is very useful to track the bottom of the first plateau. In my experience you can nail down most of the issues by just studying convergence to this first plateau. Adam and Rmsprop tend to be slower to converge and can obfuscate the true convergence rate. Just my experience, no hard science here.

**Comment by revilokeb** [comment]
Cool thanks for sharing your experience, very helpful even if it is no hard science :-). It could well be that it makes a lot of sense to go back to something simpler than full ILSVRC12 on relatively high resolution for better understanding convergence...but for the moment I will keep going on full dataset and see how it is developing.
I am providing you with LMDB then, too, as soon as I have uploaded them. If you like you might play with it a bit, too.

**Comment by beniz** [comment]
Thanks I'll wait for your lmdbs. Depending on memory requirements, it could be of interest to have weights for v3 as well. My view is that finetuning is the most prevalent use case, with application to problems that are simpler than ilsvrc in structure. As a consequence in practice it is useful to have a good set of weights for a network than can be trained on a single, possibly 4GB, GPU.

I haven't looked in detail yet at the TF to Caffe conversion, though the opposite is readily available, so it shouldn't be too difficult unless I am missing something.

**Comment by revilokeb** [comment]
Here is my solver that I am using at the moment: https://gist.github.com/revilokeb/b6c10197898662ed83832ee871da4528

And here the train_val: https://gist.github.com/revilokeb/ab1809954f69d6d707be0c301947b69e

I have started uploading LMDB with images resized to 328x328 to some cloud folder. I will also put another LMDB with images resized 128x128 next to it (which I have prepared but had no time to use yet). The upload will certainly take a while. I would make the stuff available to your email in the deepdetect rep. 

Snapshots (every 10 epochs) I will put here: https://drive.google.com/open?id=0B1qLpHDbczM2bTZyeHc0TmNoanM

I agree that finetuning is the most prevalent use case, and being able to finetune on single 4-6GB GPU is very useful. I havent tried it of course but I would guess that inception-resnet-v2 should be possible to finetune on single 6GB, maybe even 4GB (based on the fact that mini batches of size 17 for 299x299 images just fit onto a single 12GB). 

Yes, weights for inception v3 would also be good to have, probably not too difficult as you say but I havent looked at it either.

**Comment by revilokeb** [comment]
Anyone who is interested in my efforts to train inception-resnet-v2 architecture on ILSVRC12 from scratch might check https://github.com/revilokeb/inception_resnetv2_caffe where I plan posting error / loss and snapshots every 10 epochs.

**Comment by beniz** [comment]
Thanks, you may be interested in this as well, http://arxiv.org/abs/1604.04112 (if you haven't read it all already :) )

**Comment by revilokeb** [comment]
Many thanks, one more motivation to look more closely at ELUs. Maybe as a next step... :-)

**Comment by beniz** [comment]
Very interesting novel wide residual networks, https://github.com/szagoruyko/wide-residual-networks that should be considered for integration. They appear simpler and easier to train than resnets, for similar or better results.

**Comment by revilokeb** [comment]
Whow! Many thanks for sharing, this is looking extremely interesting indeed and I will definitely check their torch code. Excited about their upcoming imagenet results, too.

Btw inception-resnet-v2 exploration is still ongoing on smaller image size incl ELU, but even there proves to be not straightforward so far ... but not giving up yet.

**Comment by revilokeb** [comment]
I have learnt that there has been a succesful training of inception-resnet-v2 in torch (albeit not yet to the accuracy of the original publication):https://github.com/lim0606/torch-inception-resnet-v2. It seems that my likely inaccuracy has been the omission of the scale factor in the last layer. I will check that as soon as I find the time and then update.

**Comment by n3011** [comment]
a better implementation of inception v3 https://github.com/n3011/Inception_v3_GoogLeNet .

**Comment by beniz** [comment]
Current inception support:

|                          | Caffe | Tensorflow | Source        | Top-1 Accuracy (ImageNet) |                                                                                                                                                                                   
|--------------------------|-------|------------|---------------|---------------------------|                                                                                                                                                                                   
| AlexNet                  | Y     | N          | BVLC          |          57.1%                 |                                                                                                                                                                              
| SqueezeNet               | Y     | N          | DeepScale              |       59.5%                    |                                                                                                                                                                     
| Inception v1 / GoogleNet | [Y](https://deepdetect.com/models/ggnet/bvlc_googlenet.caffemodel)     | [Y](https://deepdetect.com/models/tf/inception_v1.pb)          | BVLC / Google |             67.9%              |                                                         
| Inception v2             | N     | [Y](https://deepdetect.com/models/tf/inception_v2.pb)          | Google        |     72.2%                      |                                                                                                                          
| Inception v3             | N     | [Y](https://deepdetect.com/models/tf/inception_v3.pb)          | Google        |         76.9%                  |                                                                                                                          
| ResNet 50                | [Y](https://deepdetect.com/models/resnet/ResNet-50-model.caffemodel)     | [Y](https://deepdetect.com/models/tf/resnet_v1_50/resnet_v1_50.pb)          | MSR           |      75.3%                     |                                          
| ResNet 101               | [Y](https://deepdetect.com/models/resnet/ResNet-101-model.caffemodel)     | [Y](https://deepdetect.com/models/tf/resnet_v1_101/resnet_v1_101.pb)          | MSR           |        76.4%                   |                                       
| ResNet 152               | [Y](https://deepdetect.com/models/resnet/ResNet-152-model.caffemodel)     | [Y](https://deepdetect.com/models/tf/resnet_v1_50/resnet_v1_152.pb)         | MSR           |               77%            |                                           
| Inception-ResNet-v2      | N     | [Y](https://deepdetect.com/models/tf/inception_resnet_v2.pb)          | Google        |       79.79%                    |                                                                                                                  
| VGG-16                   | [Y](https://deepdetect.com/models/vgg_16/VGG_ILSVRC_16_layers.caffemodel)     | [Y](https://deepdetect.com/models/tf/vgg_16/vgg_16.pb)          | Oxford        |               70.5%            |                                                 
| VGG-19                   | [Y](https://deepdetect.com/models/vgg_16/VGG_ILSVRC_16_layers.caffemodel)     | [Y](https://deepdetect.com/models/tf/vgg_19/vgg_19.pb)          | Oxford        |               71.3%            |
